clang/lib/3C/ProgramInfo.cpp

```diff
@@ -403,12 +405,38 @@ ProgramInfo::insertIntoExternalFunctionMap(ExternalFunctionMapType &Map,
         newC->brainTransplant(oldC, *this);
         Map[FuncName] = newC;
         RetVal = true;
-      } else
+      } else {
         // if the current FV constraint is not a definition?
         // then merge.
-        oldC->mergeDeclaration(newC, *this);
-    } else if (newC->hasBody())
-      llvm::errs() << "Warning: duplicate definition " << FuncName << "\n";
+        std::string BaseFailed = "";
+        std::string ReasonFailed = "";
+        oldC->mergeDeclaration(newC, *this, ReasonFailed);
+        bool MergingFailed = BaseFailed != ReasonFailed;
+        if (MergingFailed) {
+          clang::DiagnosticsEngine &DE = C->getDiagnostics();
+          unsigned MergeFailID = DE.getCustomDiagID(DiagnosticsEngine::Fatal,
+                                                    "merging failed for %q0 due to %1");
+          const auto Pointer = reinterpret_cast<intptr_t>(FD);
+          const auto Kind = clang::DiagnosticsEngine::ArgumentKind::ak_nameddecl;
+          auto DiagBuilder = DE.Report(FD->getLocation(), MergeFailID);
+          DiagBuilder.AddTaggedVal(Pointer, Kind);
+          DiagBuilder.AddString(ReasonFailed);
+        }
+        if(MergingFailed) {
+          // Kill the process and stop conversion
+          // Without this code here, 3C simply ignores this pair of functions
+          // and converts the rest of the files as it will (in semi-compliance
+          // with Mike's (2) listed on the original issue (#283)
+          exit(1);
```

Comment:
FYI, I ran across [this page](https://llvm.org/docs/ProgrammersManual.html#id23) with `report_fatal_error` or `ExitOnError`, which might be more idiomatic, but I don't think it's important.

Reply:
That's interesting. I find the `ExitOnError` explanation strangely hard to comprehend. Is it saying that instead of reporting a diagnostic and then calling `exit`, you should declare an `ExitOnError` class and then somehow invoke it when you have a fatal error? What about other warnings/errors that you might emit? We are now using [clang::DiagnosticsEngine](http://www.goldsborough.me/c++/clang/llvm/tools/2017/02/24/00-00-06-emitting_diagnostics_and_fixithints_in_clang_tools/) but I don't see that mentioned here. Your link is to LLVM stuff, not clang stuff, so perhaps that's the reason.

Reply:
In general, LLVM code is supposed to recover after issuing a diagnostic.  This is a special case: I think we would report the diagnostic about the conflicting declaration (with the same message we envision from an enhanced version of 3C capable of recovering from this situation) and _then_ exit with a message stating that such recovery is not (yet) implemented.  Here we would use `report_fatal_error`.  `ExitOnError` can be used to make code more concise when calling a function that is already designed to return an LLVM `Error` object containing an appropriate message.

> Your link is to LLVM stuff, not clang stuff

My understanding is that guides like this apply to the whole LLVM umbrella, including Clang.  In fact, this guide contains one mention of Clang.

Again, none of this is important for this PR, but it may be worth knowing.